packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js

```diff
@@ -0,0 +1,463 @@
+  handleClose: PropTypes.func.isRequired,
+};
+
+const isLastBlockImageOrCode = (editor) => {
+  const { selection } = editor;
+
+  if (!selection) return false;
+
+  const [currentImageORCodeBlock] = Editor.nodes(editor, {
+    at: selection,
+    match: (n) => n.type === 'image' || n.type === 'code',
+  });
+
+  if (currentImageORCodeBlock) {
+    const [, currentNodePath] = currentImageORCodeBlock;
+
+    const isNodeAfter = Boolean(Editor.after(editor, currentNodePath));
+
+    return !isNodeAfter;
+  }
+
+  return false;
+};
+
+export const BlocksDropdown = () => {
+  const editor = useSlate();
+  const { formatMessage } = useIntl();
+  const [isMediaLibraryVisible, setIsMediaLibraryVisible] = React.useState(false);
+
+  const blocks = useBlocksStore();
+  const blockKeysToInclude = Object.entries(blocks).reduce((currentKeys, entry) => {
+    const [key, block] = entry;
+
+    return block.isInBlocksSelector ? [...currentKeys, key] : currentKeys;
+  }, []);
+
+  const [blockSelected, setBlockSelected] = React.useState(Object.keys(blocks)[0]);
+
+  /**
+   * @param {string} optionKey - key of the heading selected
+   */
+  const selectOption = (optionKey) => {
+    toggleBlock(editor, blocks[optionKey].value);
+
+    setBlockSelected(optionKey);
+
+    if (isLastBlockImageOrCode(editor)) {
+      // insert blank line to add new blocks below code or image blocks
+      Transforms.insertNodes(
+        editor,
+        {
+          type: 'paragraph',
+          children: [{ type: 'text', text: '' }],
+        },
+        { at: [editor.children.length] }
+      );
+    }
+
+    if (optionKey === 'image') {
+      setIsMediaLibraryVisible(true);
+    }
+  };
+
+  return (
+    <>
+      <Select
+        startIcon={<Icon as={blocks[blockSelected].icon} />}
+        onChange={selectOption}
+        placeholder={blocks[blockSelected].label}
+        value={blockSelected}
+        aria-label={formatMessage({
+          id: 'components.Blocks.blocks.selectBlock',
+          defaultMessage: 'Select a block',
+        })}
+      >
+        {blockKeysToInclude.map((key) => (
+          <BlockOption
+            key={key}
+            value={key}
+            label={blocks[key].label}
+            icon={blocks[key].icon}
+            matchNode={blocks[key].matchNode}
+            handleSelection={setBlockSelected}
+            blockSelected={blockSelected}
+          />
+        ))}
+      </Select>
+      {isMediaLibraryVisible && <ImageDialog handleClose={() => setIsMediaLibraryVisible(false)} />}
+    </>
+  );
+};
+
+const BlockOption = ({ value, icon, label, handleSelection, blockSelected, matchNode }) => {
+  const { formatMessage } = useIntl();
+  const editor = useSlate();
+
+  const isActive = isBlockActive(editor, matchNode);
+  const isSelected = value === blockSelected;
+
+  React.useEffect(() => {
+    if (isActive && !isSelected) {
+      handleSelection(value);
+    }
+  }, [handleSelection, isActive, isSelected, value]);
```

Comment:
How come we have to do this? The SelectOption calls `onChange` when a new option is selected? This feels like unnecessarily boilerplate? Or if it's not maybe we should have a comment explaining this

Reply:
this useEffect is to handle the case I mentioned before but you are right, I can provide a comment to explain why I added this part